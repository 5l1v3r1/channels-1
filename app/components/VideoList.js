// @flow
import React, { Component } from 'react';
import Video from './core/Video';
import Loader from './core/Loader';
import BlankState from './core/BlankState';
import ErrorState from './core/ErrorState';
import InfiniteScroll from './core/InfiniteScroll';
import VideoType from '../dataTypes/videoType';
import styles from './VideoList.scss';

export default class VideoList extends Component {
  props: {
    channelId: string,
    videos: VideoType[],
    isFetching: boolean,
    error: boolean,
    pageToken: string,
    fetchVideos: () => Promise<any>,
    updateActiveVideo: (video: VideoType) => void
  };

  componentDidMount() {
    this.props.fetchVideos();
  }

  componentWillReceiveProps(nextProps: any) {
    const id = this.props.channelId;
    const nextId = nextProps.channelId;
    if (id !== nextId) {
      this.props.fetchVideos();
    }
  }

  render() {
    const {
      videos,
      isFetching,
      error,
      pageToken,
      fetchVideos,
      updateActiveVideo
    } = this.props;
    if ((isFetching && !pageToken) || error) {
      return (
        <div className={styles.videoList}>
          {isFetching ? (
            <Loader />
          ) : (
            <ErrorState
              message={'Error requesting videos.'}
              color={'black'}
              retry={fetchVideos}
            />
          )}
        </div>
      );
    }

    return (
      <div className={styles.videoList}>
        {videos.length ? (
          <InfiniteScroll
            isFetching={isFetching}
            pageToken={pageToken}
            loadMore={fetchVideos}
          >
            <ul className={styles.videos}>
              {videos.map(video => (
                <Video
                  key={video.id}
                  title={video.title}
                  thumbnail={video.thumbnail}
                  description={video.description}
                  publishedAt={video.publishedAt}
                  viewCount={video.viewCount}
                  handleClick={() => updateActiveVideo(video)}
                />
              ))}
            </ul>
          </InfiniteScroll>
        ) : (
          <BlankState
            message={'No videos found.'}
            color={'black'}
          />
        )}
      </div>
    );
  }
}
