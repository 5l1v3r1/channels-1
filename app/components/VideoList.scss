@import '../app.global.scss';

.videoList {
  height: 100%;
  padding-top: 60px;
}

.videos {
  margin: 0;
  padding: 0;
  list-style: none;
}

.loader,
.blankState,
.errorState,
.notification {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
}

.notification {
  flex-direction: column;
}

.message {
  margin: 0;
  font-weight: 100;
  line-height: 3rem;
  text-align: center;

  @include font-size(36);

  span {
    display: block;
    margin: 0.5rem 0 1rem;
    color: $gray;
    line-height: 1.5rem;

    @include font-size(14);
  }
}

.select {
  padding: 0.25rem 1rem;
  border: 1px solid $gray;
  border-radius: 3px;
  font-weight: 700;
  background-color: $white;
  cursor: default;

  @include font-size(12);

  &:focus {
    outline: 0;
  }

  &:active {
    color: $white;
    background-color: $gray;
  }
}
