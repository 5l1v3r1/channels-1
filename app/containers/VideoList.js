import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as activeVideoActions from '../actions/activeVideo';
import * as videosActions from '../actions/videos';
import VideoList from '../components/VideoList';
import { getVideos } from '../reducers/index';

function mapStateToProps(state) {
  return {
    channelId: state.channels.activeId,
    videos: getVideos(state),
    pageToken: state.videos.pageToken,
    isFetching: state.videos.isFetching,
    error: state.videos.error
  };
}

const actions = {
  ...activeVideoActions,
  ...videosActions
};

function mapDispatchToProps(dispatch) {
  return bindActionCreators(actions, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(VideoList);
