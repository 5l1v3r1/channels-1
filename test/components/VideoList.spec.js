import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import VideoList from '../../app/components/VideoList';

function setup() {
  const props = {
    activeChannel: undefined,
    videos: [],
    isFetching: false,
    error: false,
    pageToken: '',
    fetchVideos: jest.fn(),
    toggleChannelList: jest.fn()
  };

  const enzymeWrapper = shallow(<VideoList {...props} />);

  return {
    props,
    enzymeWrapper
  };
}

describe('VideoList', () => {
  describe('with no active channel', () => {
    it('should render self with notification', () => {
      const { enzymeWrapper } = setup();
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });
  });

  describe('with active channel', () => {

    const { props, enzymeWrapper } = setup();
    beforeEach(() => {
      enzymeWrapper.setProps({
        activeChannel: {
          id: 'UCyIe-61Y8C4_o-zZCtO4ETQ',
          subscriberCount: '243,643',
          thumbnail: 'https://yt3.ggpht.com/photo.jpg',
          title: 'DevTips',
          videoCount: '292'
        }
      });
    });

    afterEach(() => {
      enzymeWrapper.setProps({
        ...props
      });
    });

    it('should render self with loader', () => {
      enzymeWrapper.setProps({
        isFetching: true
      });
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });

    it('should render self with error message', () => {
      enzymeWrapper.setProps({
        error: true
      });
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });

    it('should render self with videos', () => {
      enzymeWrapper.setProps({
        videos: [
          {
            description: 'I visited Adobe Max and asked all my design heroes for advice.',
            id: '4y_G6BayFz0',
            publishedAt: 'a year ago',
            thumbnail: 'https://i.ytimg.com/hqdefault.jpg',
            title: 'Professional Advice from Adobe MAX Conference',
            viewCount: '5,463'
          },
          {
            description: 'Wired Article',
            id: 'Flze-rwT7lM',
            publishedAt: '7 months ago',
            thumbnail: 'https://i.ytimg.com/vi/Flze-rwT7lM/hqdefault.jpg',
            title: 'If You Want to Survive in Design, You Better Learn to Code',
            viewCount: '39,430'
          }
        ]
      });
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });

    it('should render self with no videos', () => {
      const tree = toJson(enzymeWrapper);

      expect(tree).toMatchSnapshot();
    });
  });

  it('should call toggleChannelList', () => {
    const { props, enzymeWrapper } = setup();
    const select = enzymeWrapper.find('.select');

    select.simulate('click');

    expect(props.toggleChannelList).toBeCalled();
  });
});
