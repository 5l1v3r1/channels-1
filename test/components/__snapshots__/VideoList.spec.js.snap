// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoList with active channel should render self with error message 1`] = `
<div
  className="videoList"
>
  <ErrorState
    message="Error requesting videos."
    retry={[Function]}
  />
</div>
`;

exports[`VideoList with active channel should render self with loader 1`] = `
<div
  className="videoList"
>
  <Loader />
</div>
`;

exports[`VideoList with active channel should render self with no videos 1`] = `
<div
  className="videoList"
>
  <BlankState
    message="No videos found."
  />
</div>
`;

exports[`VideoList with active channel should render self with videos 1`] = `
<div
  className="videoList"
>
  <InfiniteScroll
    isFetching={false}
    loadMore={[Function]}
    pageToken=""
  >
    <ul
      className="videos"
    >
      <Video
        description="I visited Adobe Max and asked all my design heroes for advice."
        publishedAt="a year ago"
        thumbnail="https://i.ytimg.com/hqdefault.jpg"
        title="Professional Advice from Adobe MAX Conference"
        viewCount="5,463"
      />
      <Video
        description="Wired Article"
        publishedAt="7 months ago"
        thumbnail="https://i.ytimg.com/vi/Flze-rwT7lM/hqdefault.jpg"
        title="If You Want to Survive in Design, You Better Learn to Code"
        viewCount="39,430"
      />
    </ul>
  </InfiniteScroll>
</div>
`;

exports[`VideoList with no active channel should render self with notification 1`] = `
<div
  className="videoList"
>
  <div
    className="notification"
  >
    <p
      className="message"
    >
      No Active Channel
      <span>
        Well don't just sit there...
      </span>
    </p>
    <button
      className="select"
      onClick={[Function]}
    >
      Select a Channel
    </button>
  </div>
</div>
`;
